File: exercises/chapter2.cpp
#include <RayTracerChallenge/objects/canvas.hpp>
#include <RayTracerChallenge/datastructures/vec4.hpp>
#include <fstream>
#include <iostream>

class Projectile {
public:
  rtc::Vec4 pointPosition;
  rtc::Vec4 vecVelocity;

  Projectile(const rtc::Vec4 &pos, const rtc::Vec4 &vel) : pointPosition{pos},
    vecVelocity{vel} {
  }
};

class Environment {
public:
  rtc::Vec4 vecGravity;
  rtc::Vec4 vecWind;

  Environment(const rtc::Vec4 &grav, const rtc::Vec4 &wind) : vecGravity{grav},
    vecWind{wind} {
  }
};

void tick(const Environment &env, Projectile &proj) {
  const auto point = proj.pointPosition + proj.vecVelocity;
  const auto vel = proj.vecVelocity + env.vecGravity + env.vecWind;
  proj.pointPosition = point;
  proj.vecVelocity = vel;
}

int main() {
  std::cout << "Hello" << '\n';
  Projectile proj{rtc::point(0, 1, 0),
                  rtc::vector(1, 1.8, 0).normalise() * 10.8};
  const Environment env{rtc::vector(0, -0.1, 0), rtc::vector(-0.01, 0, 0)};
  rtc::Canvas c{900, 500};
  const rtc::Colour red{1, 0, 0};

  while (proj.pointPosition.y > 0.000001) {
    c.writePixel(static_cast<std::size_t>(proj.pointPosition.x),
                 c.getHeight() - static_cast<std::size_t>(proj.pointPosition.y),
                 red);
    tick(env, proj);
  }

  std::ofstream file;
  file.open("Chapter2.ppm");
  file << c.toPPM();
  file.close();

  return 0;
}

File: RTC.Console/Exercises/Chapter2.cs
using RTC.Console.Exercises.Structs;
using RTC.Datastructures;
using RTC.Objects;
using Environment = RTC.Console.Exercises.Structs.Environment;

namespace RTC.Console.Exercises;

public class Chapter2 : IChapterExercise
{
    public void Run()
    {
        var proj = new Projectile
        {
            Position = Vec4.Point(0, 1, 0),
            Velocity = Vec4.Vector(1, 1.8, 0).Normalised * 10.8
        };
        var env = new Environment
        {
            Gravity = Vec4.Vector(0, -0.1, 0),
            Wind = Vec4.Vector(-0.01, 0, 0)
        };
        var canvas = new Canvas(960, 540);
        Colour red = new(1, 0, 0);

        while (proj.Position.Y > 0.00001)
        {
            if (!canvas.WritePixel((int)proj.Position.X, canvas.Height - (int)proj.Position.Y, red))
                System.Console.WriteLine($"Out of bounds: ({proj.Position.X}, {proj.Position.Y})");
            proj = Tick(env, proj);
        }

        canvas.SavePpm("ChapterTwo.ppm");
        canvas.SavePng("ChapterTwo.png");
    }

    private static Projectile Tick(Environment env, Projectile proj)
    {
        return new Projectile
        {
            Position = proj.Position + proj.Velocity,
            Velocity = proj.Velocity + env.Gravity + env.Wind
        };
    }
}